### PR TITLE
Make areEqualEigenvectors part of VectorX.

### LINKED ISSUE #522: Improve API for areEqualEigenvectors
**Describe the current state of a part of the code**
`areEqualEigenvectors()` is currently a stand-alone function.

**Describe a change that would improve the code base**
Since the entry point for the API is unambiguously an 'eigenvector', this method should become a member of `GQCP::VectorX`.

### PR DESCRIPTION
**Pull request by dariatols**
**Short description**
Move functions such as `areEqualEigenvectors`, `areEqualEigenvalues` and `areEqualSetsOfEigenvectors` to `VectorX`.

**Related issues**
Closes #522.

### CODE REVIEW

**Comment by dariatols** [comment]
To do:
- [ ] `areEqualEigenvalues` 
- [ ] `areEqualSetsOfEigenvectors`

File: gqcp/include/Mathematical/Representation/Matrix.hpp [hunk 1]
```diff
@@ -169,6 +169,64 @@ class Matrix:
...
+    }
+
+
+    /**
+    *  @param eigenvector1     the first eigenvector
+    *  @param eigenvector2     the second eigenvector
+    *  @param tolerance        the tolerance for comparison
+    *
+    *  @return if two eigenvectors are equal within a given tolerance
+    */
```
Comment (lelemmen) [inline gqcp/include/Mathematical/Representation/Matrix.hpp:190]: Maybe as a small guideline for better generated documentation:

```suggestion
    /**
     *  Check if two eigenvectors are equal, within a given tolerance.
     *
     *  @param eigenvector1     The first eigenvector.
     *  @param eigenvector2     The second eigenvector.
     *  @param tolerance        The tolerance for comparison.
     *
     *  @return If two eigenvectors are equal.
     */
```

File: gqcp/tests/Mathematical/Representation/Matrix_test.cpp [hunk 2]
```diff
@@ -91,6 +91,73 @@ BOOST_AUTO_TEST_CASE(Matrix_FromFile) {
...
+    // Test areEqualEigenvectors with an example.
+    GQCP::VectorX<double> a {3};
+    GQCP::VectorX<double> b {3};
+    GQCP::VectorX<double> c {3};
+    GQCP::VectorX<double> d {3};
+
+    a << 2, 3, 1;
+    b << 2, 3, 1;
+    c << -2, -3, -1;
+    d << 2, 3, 0;
```
Comment (lelemmen) [inline gqcp/tests/Mathematical/Representation/Matrix_test.cpp:105]: If I remember correctly, you can use the brace initializer here.

```suggestion
    GQCP::VectorX<double> a {2, 3, 1};
    GQCP::VectorX<double> b {2, 3, 1};
    GQCP::VectorX<double> c {-2, -3, -1};
    GQCP::VectorX<double> d {2, 3, 0};
```

Maybe you'll have to specify the type `GQCP::Vector<double, 3>`.

**Review by lelemmen — COMMENTED** [review]
As we discussed, we can even try to make these functions member functions of `Matrix` (keeping in mind some `enable_if`s).

I've also added an example of how the generated documentation can look better.

Thanks for this work, it looks good!

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 3]
```diff
@@ -1490,7 +1489,7 @@ class LinearExpansion {
             return false;
         }
 
-        return areEqualEigenvectors(this->coefficients(), other.coefficients(), tolerance);
+        return (this->coefficients()).isEqualEigenvectorAs(other.coefficients(), tolerance);
```
Comment (lelemmen) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1492]: I like the API change. For me, this reads pretty naturally.

File: gqcp/tests/Mathematical/Optimization/Eigenproblem/EigenproblemSolver_test.cpp [hunk 4]
```diff
@@ -228,7 +228,11 @@ BOOST_AUTO_TEST_CASE(Davidson_Liu_50_number_of_requested_eigenpairs) {
 
     for (size_t i = 0; i < number_of_requested_eigenpairs; i++) {
         BOOST_CHECK(std::abs(davidson_environment.eigenvalues(i) - ref_lowest_eigenvalues(i)));
-        BOOST_CHECK(GQCP::areEqualEigenvectors(davidson_environment.eigenvectors.col(i), ref_lowest_eigenvectors.col(i), 1.0e-08));
+
+        const GQCP::VectorX<double> davidson_eigenvector = davidson_environment.eigenvectors.col(i);
+        const GQCP::VectorX<double> ref_eigenvector = ref_lowest_eigenvectors.col(i);
+        BOOST_CHECK(davidson_eigenvector.isEqualEigenvectorAs(ref_eigenvector, 1.0e-08));
```
Comment (lelemmen) [inline gqcp/tests/Mathematical/Optimization/Eigenproblem/EigenproblemSolver_test.cpp:234]: Thanks for adding these extra lines, it's indeed clearer now.

**Review by lelemmen — APPROVED** [review]
Good to go. Nice implementation!

**Pull request merged**